### PR TITLE
fix: add duplicate template endpoint

### PR DESCRIPTION
**Pull request by wanlingt**
## Problem

Previously, use-template used the duplicate form endpoint to create a copy of a template form. Admins without Read permissions were not able to successfully create a copy of the template form.

## Solution

Continuation of #5377 and #5443

- Create a separate endpoint for duplication of template forms
- Create a separate `DupeFormTemplateWizardProvider`
- Refactor `DupeFormWizardProvider` to remove `isTemplate` prop

## Tests

- [x] As a logged in admin, visit the use-template endpoint of a public form `admin/form/:formId/use-template`. Click on the button 'Use this template'. The form title should be prefixed with '[Template]' and you should be able to duplicate the form successfully.
- [x] As a logged in admin, visit the use-template endpoint of a **public** form that you're not the owner of. You should be able to create a duplicate of the form template successfully.
- [x] As a logged in admin, visit the use-template endpoint of a **private** form that you're not the owner of. You should see the admin forbidden error page.
- [x] As a public user, visit the use-template endpoint of a public form. You should be redirected to the login page.

### CODE REVIEW

File: frontend/src/features/admin-form/common/components/PreviewFormBanner/PreviewFormBanner.tsx [hunk 1]
```diff
@@ -127,12 +128,21 @@ export const PreviewFormBanner = ({
...
+          <DuplicateFormTemplateModal
+            formId={formId}
+            isTemplate
+            isOpen={isModalOpen}
+            onClose={onModalClose}
+          />
+        ) : (
+          <DuplicateFormModal
+            formId={formId}
+            isTemplate
```
Comment (mantariksh) [inline frontend/src/features/admin-form/common/components/PreviewFormBanner/PreviewFormBanner.tsx]: we don't need to pass this here anymore right? and we can remove it as a prop from `DuplicateFormModal`

File: frontend/src/features/admin-form/common/components/PreviewFormBanner/PreviewFormBanner.tsx [hunk 2]
```diff
@@ -127,12 +128,21 @@ export const PreviewFormBanner = ({
...
           )}
         </Flex>
-        <DuplicateFormModal
-          formId={formId}
-          isTemplate
-          isOpen={isModalOpen}
-          onClose={onModalClose}
-        />
+        {isTemplate ? (
+          <DuplicateFormTemplateModal
```
Comment (mantariksh) [inline frontend/src/features/admin-form/common/components/PreviewFormBanner/PreviewFormBanner.tsx]: nit on naming: seems quite similar to `DuplicateFormModal`. maybe we can mirror the endpoint terminology and call it `UseTemplateModal`?

File: frontend/src/features/admin-form/template/DuplicateFormTemplateModal/DupeFormTemplateWizardProvider.tsx [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+import {
+  CreateFormFlowStates,
+  CreateFormWizardContext,
+  CreateFormWizardContextReturn,
+} from '~features/workspace/components/CreateFormModal/CreateFormWizardContext'
+import { useCommonFormWizardProvider } from '~features/workspace/components/CreateFormModal/CreateFormWizardProvider'
+
+import { useDuplicateFormTemplateMutations } from '../mutation'
+
+export const useDupeFormTemplateWizardContext = (
```
Comment (mantariksh) [inline frontend/src/features/admin-form/template/DuplicateFormTemplateModal/DupeFormTemplateWizardProvider.tsx]: similarly we should keep the terminology ("duplicate form template" vs "use template") consistent. suggest `useUseTemplateWizardContext`. `useUse` is weird but at least avoids confusion with "duplicate form" which is a different action

File: frontend/src/features/admin-form/template/DuplicateFormTemplateModal/DupeFormTemplateWizardProvider.tsx [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    isTemplateFormLoading,
+    templateFormData?.form.title,
+  ])
+
+  const { handleSubmit } = formMethods
+
+  const {
+    dupeEmailModeFormTemplateMutation,
+    dupeStorageModeFormTemplateMutation,
+  } = useDuplicateFormTemplateMutations()
```
Comment (mantariksh) [inline frontend/src/features/admin-form/template/DuplicateFormTemplateModal/DupeFormTemplateWizardProvider.tsx]: same comment on naming

**Review by mantariksh — COMMENTED** [review]
some initial nits

File: frontend/src/features/admin-form/template/DuplicateFormTemplateModal/DupeFormTemplateWizardProvider.tsx [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        formIdToDuplicate: formId,
+        emails: inputs.emails.filter(Boolean),
+        title: inputs.title,
+        responseMode: inputs.responseMode,
+      })
+    }
+    setCurrentStep([CreateFormFlowStates.Landing, 1])
+  })
+
+  const handleBackToDetails = () => {
```
Comment (mantariksh) [inline frontend/src/features/admin-form/template/DuplicateFormTemplateModal/DupeFormTemplateWizardProvider.tsx]: seems to be unused in all the wizards

File: src/app/routes/api/v3/admin/forms/admin-forms.form.routes.ts [hunk 6]
```diff
@@ -77,6 +77,24 @@ AdminFormsFormRouter.post(
...
+ * @returns 400 when Joi validation fails
+ * @returns 401 when user does not exist in session
+ * @returns 403 when form is private
+ * @returns 404 when form cannot be found
+ * @returns 410 when form is archived
+ * @returns 422 when user in session cannot be retrieved from the database
+ * @returns 500 when database error occurs
+ */
+AdminFormsFormRouter.post(
+  '/:formId([a-fA-F0-9]{24})/use-template/duplicate',
```
Comment (mantariksh) [inline src/app/routes/api/v3/admin/forms/admin-forms.form.routes.ts]: why nest the `duplicate` action under `use-template` in this route? "duplicate" is not really a sub-action under `use-template` right?

wdyt about just `POST /api/v3/admin/forms/:formId/use-template`?

**Review by mantariksh — COMMENTED** [review]
lgtm otherwise, but I think we should confirm the route before approval because it's hard to take back once it's out

File: src/app/routes/api/v3/admin/forms/admin-forms.form.routes.ts (hunk 6 above)
Comment (karrui) [reply src/app/routes/api/v3/admin/forms/admin-forms.form.routes.ts]: agree. can also just be `/template` imho hahahahaha....

Comment (wanlingt) [reply src/app/routes/api/v3/admin/forms/admin-forms.form.routes.ts]: i'm gonna go with `POST /api/v3/admin/forms/:formId/use-template`, the routes specify actions (e.g. `transfer-owner`, `duplicate`) so thought that `use-template` would be better than `template`

File: frontend/src/features/admin-form/template/DuplicateFormTemplateModal/DupeFormTemplateWizardProvider.tsx (hunk 4 above)
Comment (wanlingt) [reply frontend/src/features/admin-form/template/DuplicateFormTemplateModal/DupeFormTemplateWizardProvider.tsx]: naming changes made here! [03f35be](https://github.com/opengovsg/FormSG/pull/5446/commits/03f35be02c47d5c26e1fd9cb6cdd0cc2c0895807)

File: frontend/src/features/admin-form/template/DuplicateFormTemplateModal/DupeFormTemplateWizardProvider.tsx (hunk 5 above)
Comment (wanlingt) [reply frontend/src/features/admin-form/template/DuplicateFormTemplateModal/DupeFormTemplateWizardProvider.tsx]: you are right [e41c519](https://github.com/opengovsg/FormSG/pull/5446/commits/e41c519eb18bff147bc9c86f9a481392cb91d123)

File: src/app/routes/api/v3/admin/forms/admin-forms.form.routes.ts (hunk 6 above)
Comment (wanlingt) [reply src/app/routes/api/v3/admin/forms/admin-forms.form.routes.ts]: [0f0317a](https://github.com/opengovsg/FormSG/pull/5446/commits/0f0317ad68306afee07de0ef30a00fe805a39b10)

**Review by karrui — APPROVED** [review]
lgtm

**Comment by wanlingt** [comment]
> lgtm

i thought u say code smell

**Pull request merged**